src/chain_access.rs

```diff
@@ -35,146 +45,156 @@ impl LipaChainAccess {
     }
 
     pub(crate) fn sync(&mut self, confirm: &ConfirmWrapper) -> LipaResult<()> {
-        let mut cur_tip = self.esplora.get_tip_hash()?;
-
-        while self.synced_tip != cur_tip {
-            debug!("Currently synced up to block hash: {}", self.synced_tip);
-            debug!(
-                "Current blockchain tip hash: {} -> Start syncing blockchain",
-                cur_tip
-            );
-
-            self.sync_to_tip(confirm, &cur_tip)?;
-            self.synced_tip = cur_tip;
-            cur_tip = self.esplora.get_tip_hash()?;
+        loop {
+            let tip = self.esplora.get_tip_hash()?;
+            let something_new_to_watch = if let Some(data) = self.filter.drain() {
+                self.watched_txs
+                    .extend(data.txs.into_iter().map(|(txid, _)| txid));
+                self.watched_outputs.extend(data.outputs.into_iter());
+                true
+            } else {
+                false
+            };
+
+            if tip != self.synced_tip || something_new_to_watch {
+                debug!("Syncing to: {} ...", tip);
+                match self.try_sync_to(tip, confirm)? {
+                    SyncResult::Success => {
+                        self.synced_tip = tip;
+                        debug!("Synced to: {}", tip);
+                    }
+                    SyncResult::PossibleReorg => {
+                        debug!("Possible reorg detected, sync attempt was aborted");
+                    }
+                };
+            } else {
+                break;
+            }
         }
 
         Ok(())
     }
 
-    fn sync_to_tip(&mut self, confirm: &ConfirmWrapper, cur_tip: &BlockHash) -> LipaResult<()> {
-        self.inform_about_new_block(confirm, cur_tip)?;
-
-        // todo: we need to make sure that the following synchronization is only done up to cur_tip.
-        //       otherwise, we might end up in a situation where a block is being mined while while we're syncing
-        //       which would make corrupt the order of the confirmed transactions as required by the Confirm trait:
-        //       https://github.com/lightningdevkit/rust-lightning/blob/d6321e6e11b3e1b11e26617d3bab0b8c21da0b5b/lightning/src/chain/mod.rs#L128
+    fn try_sync_to(
+        &mut self,
+        tip_hash: BlockHash,
+        confirm: &ConfirmWrapper,
+    ) -> LipaResult<SyncResult> {
+        let tip;
+        if let Some(tip_val) = self.get_block_info(tip_hash)? {
```

Comment:
An `if let` statement that also has an `else` clause => Looks to me like a `match` statement would be a better fit